File: [CCS LABS] Windows Forms Web Image Crawler and Downloader Sample/[VB]-[CCS LABS] Windows Forms Web Image Crawler and Downloader Sample/VB/WebImageCrawler.ConvertedToVBNet/CookieWebClient.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Net

Class CookieWebClient
	Inherits WebClient
	Public Sub New(container As CookieContainer)
		Me.container = container
	End Sub

	Private ReadOnly container As New CookieContainer()

	Protected Overrides Function GetWebRequest(address As Uri) As WebRequest
		Dim r As WebRequest = MyBase.GetWebRequest(address)
		Dim request = TryCast(r, HttpWebRequest)
		If request IsNot Nothing Then
			request.CookieContainer = container
		End If
		Return r
	End Function

	Protected Overrides Function GetWebResponse(request As WebRequest, result As IAsyncResult) As WebResponse
		Dim response As WebResponse = MyBase.GetWebResponse(request, result)
		ReadCookies(response)
		Return response
	End Function

	Protected Overrides Function GetWebResponse(request As WebRequest) As WebResponse
		Dim response As WebResponse = MyBase.GetWebResponse(request)
		ReadCookies(response)
		Return response
	End Function

	Private Sub ReadCookies(r As WebResponse)
		Dim response = TryCast(r, HttpWebResponse)
		If response IsNot Nothing Then
			Dim cookies As CookieCollection = response.Cookies
			container.Add(cookies)
		End If
	End Sub
End Class

File: [CCS LABS] Windows Forms Web Image Crawler and Downloader Sample/[C#]-[CCS LABS] Windows Forms Web Image Crawler and Downloader Sample/C#/WebImageCrawler/CookieWebClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;

namespace WebImageCrawler
{
    class CookieWebClient : WebClient
    {
        public CookieWebClient(CookieContainer container)
    {
        this.container = container;
    }

    private readonly CookieContainer container = new CookieContainer();

    protected override WebRequest GetWebRequest(Uri address)
    {
        WebRequest r = base.GetWebRequest(address);
        var request = r as HttpWebRequest;
        if (request != null)
        {
            request.CookieContainer = container;
        }
        return r;
    }

    protected override WebResponse GetWebResponse(WebRequest request, IAsyncResult result)
    {
        WebResponse response = base.GetWebResponse(request, result);
        ReadCookies(response);
        return response;
    }

    protected override WebResponse GetWebResponse(WebRequest request)
    {
        WebResponse response = base.GetWebResponse(request);
        ReadCookies(response);
        return response;
    }

    private void ReadCookies(WebResponse r)
    {
        var response = r as HttpWebResponse;
        if (response != null)
        {
            CookieCollection cookies = response.Cookies;
            container.Add(cookies);
        }
    }
    }
}
